### ISSUE TITLE
feat: block header hash computation 🦀

### ISSUE THREAD

**Issue by EvolveArt** [keep-starknet-strange/madara#128]
# Feature Request

**Describe the Feature Request**

The block hash should be computed correctly following the starknet specs defined [here](https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/header/#block_hash).

You can find the code to change under `primitives/starknet/src/starknet_block/header.rs`

**Acceptance Criteria**

- Hash function used should be generic such that we can easily switch
- Tests should be added

**Comment by dangush** [comment]
Hi there, I'd like to work on this!

**Comment by abdelhamidbakhta** [comment]
> Hi there, I'd like to work on this!

let's go! I assigned it to you

**Comment by dangush** [comment]
Hi there, just want to make sure I have the right approach. What is the case in which different hash functions would be used for producing block hashes? 

I see that there's a Hasher trait already implemented, should the header hash function accept a function implementing this trait?

**Comment by EvolveArt** [comment]
> Hi there, just want to make sure I have the right approach. What is the case in which different hash functions would be used for producing block hashes? I'm noticing that the starknet pallet `SystemHash` is set to be `PedersenHasher`. Would this function be used for block headers too then?
> 
> Asking these questions to better understand how to make the hash function generic- should the header hash function signature be changed to include a function implementing trait Hasher as a parameter? Or should it be using whatever function is configured in the starknet pallet as the `SystemHash`?

Hello! Sorry for the late reply @dangush. Yes exactly the `SystemHash` should be used at least in a first step. In the future we might have two different hashers to comply with the current starknet specs.

And yes I believe you do have to change the function's signature to take a `Hasher` trait param too in order to make it work.

**Comment by EvolveArt** [comment]
Hey @dangush still working on this ?

**Comment by tdelabro** [comment]
@ClementWalter will take this issue in order to fix #381

**Comment by ClementWalter** [comment]
I think that we also need to fix:

```
fn get_storage_at(address: ContractAddressWrapper, key: StorageKeyWrapper) -> Result<U256, DispatchError>;
fn call(address: ContractAddressWrapper, function_selector: H256, calldata: Vec<U256>) -> Result<Vec<StarkFeltWrapper>, DispatchError>;
```

to use `FieldElement` instead of `U256` or `H256`.
